File: Completed/Go/1143.go
func longestCommonSubsequence(t1 string, t2 string) int {
    memo := make(map[frame]int)
    
    return helper(t1,t2, frame{len(t1), len(t2)}, memo)
}

type frame struct {
    l1, l2 int
}

func helper(t1, t2 string, f frame, memo map[frame]int) int {
    if val, ok := memo[f]; ok {
        return val
    }
    
    if f.l1 == 0 || f.l2 == 0 {
        return memo[f]
    } else if t1[f.l1-1] == t2[f.l2-1] {
        x := frame{f.l1-1, f.l2-1}
        memo[f] = 1+helper(t1,t2, x, memo)
    } else {
        a := frame{f.l1-1, f.l2}
        b := frame{f.l1, f.l2-1}
        
        memo[f] = max(
            helper(t1,t2, a, memo),
            helper(t1,t2, b, memo),
        )
    }
    
    return memo[f]
}

func max(a,b int) int {
    if a > b {
        return a
    }
    return b
}

File: Completed/python3/1143.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        self.memo = {}
        self.t1 = text1
        self.t2 = text2
        
        tup = (len(text1), len(text2))
        return self.helper(tup)
    
    def helper(self, tup):
        
        if tup in self.memo:
            return self.memo[tup]
        elif tup[0] == 0 or tup[1] == 0:
            return 0
        elif self.t1[tup[0]-1] == self.t2[tup[1]-1]:
            x = (tup[0]-1, tup[1]-1)
            self.memo[tup] = 1+self.helper(x)
        else:
            a = (tup[0]-1, tup[1])
            b = (tup[0], tup[1]-1)
            
            self.memo[tup] = max(
                self.helper(a),
                self.helper(b))
            
        return self.memo[tup]
